src/delta.c

```diff
@@ -38,86 +38,157 @@
+	const char *s = delta_name;
+
+	/* Ignore versions, deltas will be used based on their hashes only. */
+	for (int i = 0; i < 2; i++) {
+		s = strchr(s, '-');
+		if (!s) {
+			return false;
+		}
+		s++;
 	}
 
-	tmp = strdup(file->deltapeer->filename);
-	tmp2 = strdup(file->deltapeer->filename);
+	/* Note: SWUPD_HASH_LEN accounts for the NUL-terminator after the hash. */
+	const size_t hash_len = SWUPD_HASH_LEN - 1;
+	if (strlen(s) != (hash_len * 2 + 1)) {
+		return false;
+	}
 
-	dir = dirname(tmp);
-	base = basename(tmp2);
+	if (s[hash_len] != '-') {
+		return false;
+	}
 
-	string_or_die(&origin, "%s/%s/%s", path_prefix, dir, base);
+	hash_assign(s, from);
 
-	ret = apply_bsdiff_delta(origin, filename, deltafile);
-	if (ret) {
-		unlink_all_staged_content(file);
-		goto out;
-	}
-	xattrs_copy(origin, filename);
+	/* Consume the first hash and the separator. */
+	s += hash_len + 1;
 
-	if (!verify_file(file, filename)) {
-		unlink_all_staged_content(file);
-		goto out;
-	}
+	hash_assign(s, to);
 
-	if (xattrs_compare(origin, filename) != 0) {
-		unlink_all_staged_content(file);
-		goto out;
+	return true;
+}
+
+void apply_deltas(struct manifest *current_manifest)
+{
+	char *delta_dir;
+	string_or_die(&delta_dir, "%s/delta", state_dir);
+
+	DIR *dir = opendir(delta_dir);
+	if (!dir) {
+		/* No deltas available to apply. */
+		return;
 	}
 
-	unlink(deltafile);
+	struct dirent *ent;
+	while ((ent = readdir(dir))) {
+		if (!strcmp(ent->d_name, ".") || !strcmp(ent->d_name, "..")) {
+			continue;
+		}
+
+		char *to_staged = NULL;
+
+		char *delta_name = ent->d_name;
+		char *delta_file;
+		string_or_die(&delta_file, "%s/%s", delta_dir, delta_name);
+
+		char from[SWUPD_HASH_LEN] = {};
+		char to[SWUPD_HASH_LEN] = {};
+		if (!check_delta_filename(delta_name, &from[0], &to[0])) {
+			fprintf(stderr, "Invalid name for delta file: %s\n", delta_file);
+			goto next;
+		}
+
+		string_or_die(&to_staged, "%s/staged/%s", state_dir, to);
+
+		/* If 'to' file already exists, no need to apply delta. */
+		struct stat stat;
+		if (lstat(to_staged, &stat) == 0) {
+			goto next;
+		}
+
+		/* TODO: Sort the list of files by hash then walk it in parallel with readdir. */
+
+		struct list *ll = list_head(current_manifest->files);
+		char *found = NULL;
+
+		for (; ll && !found; ll = ll->next) {
+			struct file *file = ll->data;
+			if (file->is_deleted || file->is_ghosted || !file->is_file || !hash_equal(file->hash, from))
+				continue;
+
+			/* Verify the actual file in the disk matches our expectations. */
+			char hash[SWUPD_HASH_LEN];
+			char *filename;
+			string_or_die(&filename, "%s/%s", path_prefix, file->filename);
+
+			if (!compute_hash_from_file(filename, hash) || !hash_equal(file->hash, hash)) {
+				free_string(&filename);
+				continue;
+			}
+
+			found = filename;
+		}
+
+		if (!found) {
+			fprintf(stderr, "Couldn't use delta file %s: no 'from' file to apply was found\n", delta_file);
```

Comment:
Let's not print these warnings. The user can't do anything about them. If anything this would be something to report to telemetry, but that again would make a bunch of useless telemetry records. I'd rather just not print this kind of thing.

Reply:
This indicated a `/var/lib/swupd` on my system that had a lot of extra files hanging around. Running a `swupd clean --all` made the error go away of course, but I still don't think the prints are needed.

Reply:
I think these are good candidates for a telemetry entry, since it means we are generating packs with files that shouldn't be there. Maybe just send one telemetry entry with the count of the deltas that we couldn't use? (I'm fine just removing it too)

> This indicated a /var/lib/swupd on my system that had a lot of extra files hanging around. 

Yes, in the first execution you would get noise since I think we were not cleaning the deltas before.

Reply:
Further testing shows these errors don't show up any more. I'm fine with leaving them for now.